components/suggest/src/db.rs

```diff
@@ -128,63 +128,49 @@ impl<'a> SuggestDao<'a> {
+        let suggestions = unique_providers
             .iter()
-            .any(|p| matches!(p, SuggestionProvider::Pocket | SuggestionProvider::Amo))
-        {
-            (self.conn.prepare_cached(
-                &format!(
-                    "SELECT s.id, k.rank, s.title, s.url, s.provider, s.score, NULL as confidence, NULL as keyword_suffix
-                     FROM suggestions s
-                     JOIN keywords k ON k.suggestion_id = s.id
-                     WHERE s.provider IN ({}) AND
-                           k.keyword = :keyword
-                     UNION ALL
-                     SELECT s.id, k.rank, s.title, s.url, s.provider, s.score, k.confidence, k.keyword_suffix
-                     FROM suggestions s
-                     JOIN prefix_keywords k ON k.suggestion_id = s.id
-                     WHERE k.keyword_prefix = :keyword_prefix
-                     ORDER BY s.score DESC
-                     LIMIT :suggestions_limit",
-                    providers_to_sql_list(&query.providers),
-                ),
-            )?, vec![
-                (":keyword", keyword_lowercased as &dyn ToSql),
-                (":keyword_prefix", &keyword_prefix as &dyn ToSql),
-                (":suggestions_limit", &suggestions_limit as &dyn ToSql)
-            ])
-        } else {
-            (self.conn.prepare_cached(
-                &format!(
-                    "SELECT s.id, k.rank, s.title, s.url, s.provider, s.score, NULL as confidence, NULL as keyword_suffix
-                     FROM suggestions s
-                     JOIN keywords k ON k.suggestion_id = s.id
-                     WHERE s.provider IN ({}) AND
-                           k.keyword = :keyword
-                     ORDER BY s.score DESC
-                     LIMIT :suggestions_limit",
-                    providers_to_sql_list(&query.providers),
-                ),
-            )?, vec![
-                (":keyword", keyword_lowercased as &dyn ToSql),
-                (":suggestions_limit", &suggestions_limit as &dyn ToSql)
-            ])
-        };
+            .try_fold(vec![], |mut acc, provider| {
+                let suggestions = match provider {
+                    SuggestionProvider::Amp => self.fetch_amp_suggestions(query),
+                    SuggestionProvider::Wikipedia => self.fetch_wikipedia_suggestions(query),
+                    SuggestionProvider::Amo => self.fetch_amo_suggestions(query, suggestions_limit),
+                    SuggestionProvider::Pocket => {
+                        self.fetch_pocket_suggestions(query, suggestions_limit)
+                    }
+                    SuggestionProvider::Yelp => self.fetch_yelp_suggestions(query),
+                }?;
+                acc.extend(suggestions);
+                Ok(acc)
+            });
+
+        suggestions.map(|mut suggestions| {
+            suggestions.sort();
+            suggestions.truncate(suggestions_limit as usize);
```

Comment:
Doh, looks like we set `suggestion_limit` to -1 by default, which would be problematic since truncate expects a usize. Maybe we should do `let suggestion_limit = query.limit.unwrap_or(i32::MAX)` 😀